### ISSUE TITLE
Crash in telemetry / allocation tracking

### ISSUE THREAD

**Issue by jcward** [HaxeFoundation/hxcpp#333]
Reported by @madrazo over in #332, later found not to be isolated to that PR, so filing as a separate issue.

His notes: Likely thread-related, Win 10, URLLoader in Lime legacy, xml files use elements with the same name.

**Comment by jcward** [comment]
Hmm, I'm getting a crash in a multi-threaded load test (on Linux), but not in the same place as yours.

http://hastebin.com/fupajupibe.pas

line 1199 is: return stack->mTelemetry->Dump();

I'll do some more testing (and try to get my windows VM back online.)

**Comment by jcward** [comment]
Oh, one question @madrazo - I assume you had only started telemetry from your main thread? Or did you start telenetry (new hxtelemetry) from multiple threads? You said this was a lime project using URLLoader?

**Comment by madrazo** [comment]
@jcward Thanks. I am (still) using openfl/lime legacy. Tried with hxScout 0.4 and 0.5 but same result.
I now suspect that is something related with the object names retrieved from the XML. It happens when parsing a second XML that is similar to previous. If I skip this, the third one (non releated) parses correctly. If this is true i will be able to reproduce it with Haxe only or with one of your samples (https://github.com/jcward/hxtelemetry/tree/master/test/DisplayingABitmap) . Let me try modifiying the XML as well. 

Is there a way to start the telemetry only after my level has been loaded?  Thanks!

**Comment by jcward** [comment]
A reproducible testcase would be great!

Hmm, now that telemetry support is built into the stage-startup in OpenFL, you'd have to comment it out if you want to delay the startup. In `openfl/_legacy/display/Stage.hx` new(), there's a call to `Telemetry.__initialize ();` -- it should work to comment that out and make that call yourself at a later time.

**Comment by jcward** [comment]
I added a URLLoader to my test (loading an SVG every other frame), and then I saw this while testing with telemetry on:

    Main.hx:122: At frame: 240
    *** Error in `./DisplayingABitmap': malloc(): smallbin double linked list corrupted: 0x0000000004093d80 ***
    Aborted (core dumped)

GDB backtrace: http://hastebin.com/inudirekev.sm

Once I saw this error, but it seems unrelated:

    [xcb] Unknown sequence number while awaiting reply
    [xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
    [xcb] Aborting, sorry about that.
    DisplayingABitmap: ../../src/xcb_io.c:298: poll_for_response: Assertion `!xcb_xlib_threads_sequence_lost' failed.
    Aborted (core dumped)

Of course, I'm also testing on Linux as I don't have a Win dev machine at present.

I'm hoping @madrazo can provide a reproducible testcase.

**Comment by madrazo** [comment]
Now I am thinking that the issue is the large size of the data. At some time time, a realloc occurs but it is marking or setting free memory. So, related with the mutex. The strange thing is that it only crashes if hxScout is open. Otherwise it still works ok.

**Comment by jcward** [comment]
So it's using a URLLoader to load a large XML file and parse it? If hxScout is closed, hxtelemetry doesn't initialize, so the cpp Telemetry code doesn't run.

**Comment by madrazo** [comment]
Yes. Now my guess is that the iterator variable is getting collected before it has been finishing used it.

**Comment by madrazo** [comment]
Hi  @jcward @hughsando 

Here is the test case for this issue. 
https://github.com/madrazo/telemetryTest

It is quite simple, you can see the difference in:
https://github.com/madrazo/telemetryTest/commit/6f7b525708ca04d4c67ffbea05749b7e9f137307

Added "set" flags in XML so it runs "telemetry" and "legacy". Just need to run : 
"lime test project.xml windows" or
"lime test project.xml windows -DHXCPP_GC_MOVING"
or with -DHXCPP_DEBUG_LINK

We actually call Assets.getText between each parse and store this data, but this example encapsulates the issue that is in Xml.parse. With our real XMLs, we get a crash after parsing two or three 2-4 MB xml files.

It crashes both Moving and not Moving on "9" (with PR 332). 
Without PR 332 (https://github.com/HaxeFoundation/hxcpp/pull/332) it crashes with Moving on "4". 

It crashes when opening hxScout beforehand.

Thanks!

**Comment by hughsando** [comment]
I have done a few fixes in hxcpp for the moving case.
When I try this sample now, I see it crashing in allocation_data->push_back.  I think it might be that the test code does too many allocations between frames/flushes, and makes the allocation data buffer too large for the memory.
To fix this, I think the telemetry code may need to be able to send partial buffers - possibly have a thread consuming data and sending it as it becomes available, rather than waiting for frame events.
This might also make it run faster.

**Comment by hughsando** [comment]
@jcward What do you  think of this one?  Should I close it?

**Comment by jcward** [comment]
@hughsando Hmm, yes, I'm not actively working on this. I'll close it due to inactivity, with hopes of getting back to telemetry features and testing later. I've made a note in an issue in the hxtelemetry repo (link just above.)
